group-coordinator/src/main/java/org/apache/kafka/coordinator/group/assignor/GeneralUniformAssignmentBuilder.java

```diff
@@ -14,17 +14,891 @@
+                }
+            } else
+                addPartitionMovementRecord(partition, pair);
+        }
+
+        private TopicIdPartition getTheActualPartitionToBeMoved(TopicIdPartition partition, String oldOwner, String newOwner) {
+            Uuid topic = partition.topicId();
+
+            if (!partitionMovementsByTopic.containsKey(topic))
+                return partition;
+
+            // This partition was previously moved.
+            if (partitionMovementsByPartition.containsKey(partition)) {
+                assert oldOwner.equals(partitionMovementsByPartition.get(partition).dstMemberId);
+                oldOwner = partitionMovementsByPartition.get(partition).srcMemberId;
+            }
+
+            Map<MemberPair, Set<TopicIdPartition>> partitionMovementsForThisTopic = partitionMovementsByTopic.get(topic);
+            MemberPair reversePair = new MemberPair(newOwner, oldOwner);
+            if (!partitionMovementsForThisTopic.containsKey(reversePair))
+                return partition;
+
+            return partitionMovementsForThisTopic.get(reversePair).iterator().next();
+        }
+    }
+
+    /**
+     * Manages assignments to members based on their current assignment size and maximum allowed assignment size.
+     */
+    protected class AssignmentManager {
+        private final Map<String, MemberAssignmentData> membersWithAssignmentSizes = new HashMap<>();
+
+        /**
+         * Represents the assignment metadata for a member.
+         */
+        private class MemberAssignmentData {
+            final String memberId;
+            int currentAssignmentSize = 0;
+            int maxAssignmentSize;
+
+            /**
+             * Constructs a MemberAssignmentData with the given member Id.
+             *
+             * @param memberId The Id of the member.
+             */
+            MemberAssignmentData(String memberId) {
+                this.memberId = memberId;
+            }
+
+            @Override
+            public boolean equals(Object o) {
+                if (this == o) return true;
+                if (o == null || getClass() != o.getClass()) return false;
+                MemberAssignmentData that = (MemberAssignmentData) o;
+                return memberId.equals(that.memberId);
+            }
+
+            @Override
+            public int hashCode() {
+                return Objects.hash(memberId);
+            }
+
+            @Override
+            public String toString() {
+                return "MemberAssignmentData{" +
+                    "memberId='" + memberId + '\'' +
+                    ", currentAssignmentSize=" + currentAssignmentSize +
+                    ", maxAssignmentSize=" + maxAssignmentSize +
+                    '}';
+            }
+        }
+
+        /**
+         * Initializes an AssignmentManager, setting up the necessary data structures.
+         */
+        public AssignmentManager() {
+            initializeAssignmentSizes();
+        }
```

Comment:
could you help me understand why we should pass the global variables to initializeAssignmentSIzes?
I see your point about the helper method, my thought process was that I wanted it to be concise and easy to read.

Reply:
we bring in dependencies - this constructor can only be built once the members and subscribedTopicDescriber global variables are initialized. This will break if that order changes (which we probably won't change). Since the constructor has no arguments it makes it seem as if it does not depend on anything.

constructors are to initialize fields so i don't think we need a helper method

Reply:
understood thanks for the explanation!